sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionScopedRegionLevelProgress.java

```diff
@@ -0,0 +1,266 @@
+                                regionLevelProgressAsValInner.vectorSessionToken);
+                        });
+                    }
+                    // regionId maps to a hub region
+                    else {
+                        regionLevelProgressAsVal.compute(normalizedRegionRoutedTo, (normalizedRegionAsKey, regionLevelProgressAsValInner) -> {
+
+                            if (regionLevelProgressAsValInner == null) {
+                                return new RegionLevelProgress(vectorSessionToken.getLSN(), Long.MIN_VALUE, null);
+                            }
+
+                            // regionLevelProgressAsValInner.vectorSessionToken is passed
+                            // to have a session token to merge with in case normalizedRegionRoutedTo
+                            // is equal to the first preferred region in the subsequent step
+                            return new RegionLevelProgress(
+                                Math.max(regionLevelProgressAsValInner.getMaxGlobalLsnSeen(), vectorSessionToken.getLSN()),
+                                Long.MIN_VALUE,
+                                regionLevelProgressAsValInner.vectorSessionToken);
+                        });
+                    }
+
+                    // store the session token in parsed form if obtained from the firstEffectivePreferredReadableRegion (a merge is necessary to store latest progress from first preferred region)
+                    if (normalizedRegionRoutedTo.equals(firstEffectivePreferredReadableRegion)) {
+                        regionLevelProgressAsVal.compute(normalizedRegionRoutedTo, (normalizedRegionAsKey, regionLevelProgressAsValInner) -> {
+                            if (regionLevelProgressAsValInner == null) {
+                                return new RegionLevelProgress(vectorSessionToken.getLSN(), Long.MIN_VALUE, vectorSessionToken);
+                            }
+
+                            VectorSessionToken mergedBasedSessionToken = regionLevelProgressAsValInner.vectorSessionToken == null ? vectorSessionToken : (VectorSessionToken) regionLevelProgressAsValInner.vectorSessionToken.merge(vectorSessionToken);
+
+                            return new RegionLevelProgress(
+                                Math.max(regionLevelProgressAsValInner.getMaxGlobalLsnSeen(), vectorSessionToken.getLSN()),
+                                Long.MIN_VALUE,
+                                mergedBasedSessionToken);
+                        });
+                    }
+                } else {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("Region with name - {} which has no known regionId has been seen", normalizedRegionRoutedTo);
+                    }
+                }
+
+                return regionLevelProgressAsVal;
+            } catch (CosmosException e) {
+                throw new IllegalStateException(e);
```

Comment:
which part of the try catch block can throw the error? curious how the error handling is done in the whole session token flow?
Will it interrupt the request, or let it succeed with global session token in case of any computation error?

Reply:
I modified it to catch a generic `Exception` type. In `PartitionScopedRegionLevelProgress` the flow is purely parsing `ISessionToken` - so any error should ideally be surfaced to the caller without any silent handling (of course at this point even the global session token cannot be used for whatever reason).

Reply:
Removing any kind of exception chaining. Any merge error should not be handled silently.